packages/flutter/lib/src/material/carousel.dart

```diff
@@ -490,6 +627,506 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
+/// indicates three items should be visible at a time. The layout of these items
+/// would be:
+///  * First item: Extent is (1 / (1 + 6 + 1)) * viewport extent.
+///  * Second item: Extent is (6 / (1 + 6 + 1)) * viewport extent.
+///  * Third item: Extent is (1 / (1 + 6 + 1)) * viewport extent.
+///
+/// Assuming a viewport extent of 800 in the main axis and the first item is
+/// item 0, there would be three visible items with extents of 100, 600, and 100.
+/// As item 0 scrolls off-screen, the extent of item 1 smoothly decreases from 600
+/// to 100. For instance, if item 0 is 30% off-screen, item 1 should have decreased
+/// its size to 30% of the difference from 600 to 100; its extent would be
+/// 600 - 0.3 * (600 - 100). Similarly, item 2's extent would increase from 100
+/// to 600, becoming 100 + 0.3 * (600 - 100).
+///
+/// As the initially visible items change size during scrolling, item 3 enters
+/// the view to fill the remaining space. Its extent starts at a minimum of
+/// [shrinkExtent] (or 0 if [shrinkExtent] is not provided) and gradually
+/// increases to match the extent of the last visible item (100 in this example).
+class _SliverWeightedCarousel extends SliverMultiBoxAdaptorWidget {
+  const _SliverWeightedCarousel({
+    required super.delegate,
+    required this.allowFullyExpand,
+    required this.shrinkExtent,
+    required this.weights,
+  });
+
+  // Determine whether extra scroll offset should be calculate so that every
+  // item have a chance to scroll to the maximum extent.
+  //
+  // This is useful when the leading/trailing items has smaller weights than
+  // the middle items, such as [1,7], [3,2,1].
+  final bool allowFullyExpand;
+
+  // The starting extent for items when they gradually show on/off screen.
+  //
+  // This is useful to avoid a hairline shape. This value should also smaller
+  // than the last item extent to make sure a smooth transition. So in calculation,
+  // this is limited to [0, weight for the last visible item].
+  final double shrinkExtent;
+
+  // The layout arrangement.
+  //
+  // When items are laying out, each item will be arranged based on the order of
+  // the weights and the extent is based on the corresponding weight out of the
+  // sum of weights. The length of weights means how many items we can put in the
+  // view at a time.
+  final List<int> weights;
+
+  @override
+  RenderSliverFixedExtentBoxAdaptor createRenderObject(BuildContext context) {
+    final SliverMultiBoxAdaptorElement element = context as SliverMultiBoxAdaptorElement;
+    return _RenderSliverWeightedCarousel(
+      childManager: element,
+      allowFullyExpand: allowFullyExpand,
+      shrinkExtent: shrinkExtent,
+      weights: weights,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderSliverWeightedCarousel renderObject) {
+    renderObject.allowFullyExpand = allowFullyExpand;
+    renderObject.shrinkExtent = shrinkExtent;
+    renderObject.weights = weights;
```

Comment:
Slightly more conventional:

```suggestion
    renderObject
      ..allowFullyExpand = allowFullyExpand
      ..shrinkExtent = shrinkExtent
      ..weights = weights;
```